### ISSUE TITLE
Неправильная работа с адресами в браузере

### ISSUE THREAD

**Issue by nikitasA1NT** [BladeMight/Mahou#305]
При вводе текста в адресную строку браузера неправильной раскладкой и при последующей ее конвертации при помощи Mahou, конвертируется введенные символы, но не заменяются какие нужно, а заменяется выделенная подсказка браузера. Сначала подумал, что это из-за включенной конвертации выделенного текста (у меня одна и та же клавиша это делает), но после того как я ее отключил проблема осталась.
Пример:
![image](https://user-images.githubusercontent.com/57367746/103485447-d7630980-4e0f-11eb-94ec-3b201e1e0174.png)
После нажатия конвертации:
![image](https://user-images.githubusercontent.com/57367746/103485462-fa8db900-4e0f-11eb-8c0d-08b73dac3108.png)
Аналогично и с русского на английский:
![image](https://user-images.githubusercontent.com/57367746/103485491-2ad55780-4e10-11eb-9a86-5ba556059ef8.png)
![image](https://user-images.githubusercontent.com/57367746/103485513-493b5300-4e10-11eb-9f1d-69d3ce069b9b.png)

**Comment by BladeMight** [comment]
Хм. Может это в новом chrome добавили "автоподставку первого"? Из-за этого при посылании удаления символов которые ты ввёл "сначала удаляется автоподставка", а потом уже не все символы что ты вводил, а только на 1 меньше. Например: пишем: *ско* chrome добавляет: **рость интертнета**, Mahou посылает Back в первый раз, удаляется "автоподставка chrome": **рость интертнета**, а потом уже только "ко", а "с" остаётся...

 У меня на 74.0.3729.169 такого нет:
![pads](https://user-images.githubusercontent.com/11931888/103486225-56a30e00-4e0d-11eb-922e-301bc3f91200.gif)

**Comment by nikitasA1NT** [comment]
Да, я так и подумал что вы напишите, что невозможно программно определить подсказку хрома. Но до этого я пользовался Punto Switcher, но да, он не опенсорсный. Другие аналоги не пробовал, по функционалу и нужной мне фиче, меня ваша программа максимально устраивает. Может можно порыться в их исходном коде и понять как они это делают, если делают.
Вероятно можно как-то узнать, какой выделенный текст был выделен пользователем, а какой нет и от этого исходить.
Я полагаю, реализация у вас такая: программа запоминает введенные символы, после нажатия конвертации, эти символы она стирает и заменяет на аналоги с другой раскладки? А она может как-то проводить поиск в уже введенном тексте? Если да, то можно сделать поиск, тех символов, которые программа, должна конвертировать, если они были не конвертированы корректно, как в насущной проблеме.
Хотя это костыли, потому что если она будет проводить этот самый поиск и окажется так, что она уже заменила все корректные символы далее найден точно такие же, то заменит и их, которые не входили в нужную замену и были введены давно.
Снова хотя, можно завести булевую переменную, которая будет true, в случае, если программа заменила один раз нужные символы и в случае этого самого true, поиск, о котором я писал, уже не выполнять.
Короче сумбурно, я тоже не знаю, потому что надо разбираться в текущей реализации.

**Comment by BladeMight** [comment]
> Я полагаю, реализация у вас такая: программа запоминает введенные символы, после нажатия конвертации, эти символы она стирает и заменяет на аналоги с другой раскладки? 

Конвертация "ввода", а это "последнего введенного слова/линии" - работает таким образом:
1. Mahou запоминает клавиши которые ты вводил.
2. При некоторых условиях: Space, смена окон, Enter, и т.д. "запомненные клавиши могут очистится", т.е. началось новое слово/линия.
3. При нажатии на горячую клавишу конвертации последнего слова/линии, Mahou сначала посылает клавишу <kbd>Back</kbd> "точное количество введенных символов", тем самым удаляя символы которые были введены. Если на этом моменте chrome сам решит "подставить" свои символы, тогда получится что Mahou удалит их вместо "введенного символа".
4. Потом Mahou меняет раскладку.
5. Mahou посылает те же самые клавиши которые ты вводил. Т.е. если ты вводил до этого слово "ghbdtn" - это клавиши: <kbd>g</kbd><kbd>h</kbd><kbd>b</kbd><kbd>d</kbd><kbd>t</kbd><kbd>n</kbd>, и именно их Mahou и посылает.

>  то можно проверить, стерла ли она реально нужный символ.

Если бы такое было возможно было бы намного проще… Эти данные в самом Chrome, системой/WinAPI их не получить, а chrome получает только саму клавишу BackSpace, и Mahou не понятно что он удаляет.

**Comment by BladeMight** [comment]
Ну можно еще просто добавить "1 лишний" Back, т.е. чтобы затирало всегда на 1 больше, но тогда для слов посередине строки возможно будут удалятся пробелы между ними...

**Comment by BladeMight** [comment]
> Но до этого я пользовался Punto Switcher, но да, он не опенсорсный, и он умел корректно конвертировать раскладку в адресной строке браузера при его подсказках. 

@nikitasA1NT  Клавишей Break, то бишь горячей клавишей "отмены конвертации раскладки"? Сомневаюсь, вероятно ты говоришь про его авто-замену, которая срабатывает по ходу "ввода". В Mahou есть авто-замена но тут она срабатывает только при Space, и только если включена.

**Comment by nikitasA1NT** [comment]
Нет, я автозаменой не пользовался

**Comment by nikitasA1NT** [comment]
Я придумай кое-что. Есть такая комбинация клавиши как Shift+СтрелкаВлево, которая выделяет текст. Можно изменить действие программы, чтобы она посылала Back не столько раз, сколько символов было введено, а посылала Shift+СтрелкаВлево столько раз, и одну клавишу Back. Думаю это флекс идея, потому так подсказка от хрома тоже стерется, и при безо всяких подсказках других программ оно тоже работает😃👍

**Comment by nikitasA1NT** [comment]
Если у вас нет лишнего времени, то я сам бы хотел за это взяться, если вы мне объясните, где в коде находится тот самый алгоритм, который вы описывали выше

**Comment by BladeMight** [comment]
> безо всяких подсказках других программ оно тоже работает

в cmd не будет работать.

 Я придумал другой способ в e1e4ae7:　

Получать выделенный текст(Cltr+Insert) до конвертации, если он не пустой тогда добавлять 1 Back, вот:
![seledka](https://user-images.githubusercontent.com/11931888/103489589-b823a680-4e26-11eb-91cb-db069810064b.gif)

**Comment by nikitasA1NT** [comment]
Фиксирую, теперь все ок👍

**Comment by nikitasA1NT** [comment]
А никак нельзя чтобы из-за https://github.com/BladeMight/Mahou/commit/e1e4ae77185394f0ccc09d8806ca535a79bd2d24 не затирался предыдущий буфер обмена. А то сейчас при повседневке это очень раздражать начало. Скопипую что-то, а потом удивляюсь, где мое скопированное что-либо 😕

**Comment by BladeMight** [comment]
@nikitasA1NT По идее и не должно затирать буфер, в [e1e4ae7 Line#2832](https://github.com/BladeMight/Mahou/commit/e1e4ae77185394f0ccc09d8806ca535a79bd2d24#diff-cc1dc3fbbef6a8eeebd42b3cf727b698f248fc9134b8973226a35f6348f924dcR2832) я делаю backup перед "получением буфера обмена в `GetClipStr()`(посылает Ctrl+Insert/C и получает буфер) [e1e4ae7 Line#2833](https://github.com/BladeMight/Mahou/commit/e1e4ae77185394f0ccc09d8806ca535a79bd2d24#diff-cc1dc3fbbef6a8eeebd42b3cf727b698f248fc9134b8973226a35f6348f924dcR2833) и наконец после всего этого возвращаю буфер в `RestoreClipboard()` [e1e4ae7 Line#2836](https://github.com/BladeMight/Mahou/commit/e1e4ae77185394f0ccc09d8806ca535a79bd2d24#diff-cc1dc3fbbef6a8eeebd42b3cf727b698f248fc9134b8973226a35f6348f924dcR2836).

Хм... Возможно таки конфликтует с такой же конструкцией (backup=>do_things=>restore) из "Конвертации выделения", если у тебя назначена одна горячая клавиша на обе действия...

**Comment by BladeMight** [comment]
Вот:
![copy_backup](https://user-images.githubusercontent.com/11931888/103971459-af5f0780-517b-11eb-985a-8688ff44288e.gif)

А, еще вот вспомнил: другие форматы кроме "текста" будут удалены… Mahou пока не поддерживает восстановление буфера обмена если в нём что-то кроме текста. И [восстановить некоторые форматы будет крайне непросто...](http://stackoverflow.com/questions/2578900/how-do-i-backup-and-restore-the-system-clipboard-in-c).

**Comment by BladeMight** [comment]
@nikitasA1NT В 1d14c8c добавил полную поддержку восстановление буфера обмена ⭐🎉⭐ , демо с Paint'ом:
![mahou_full_clip_backup](https://user-images.githubusercontent.com/11931888/103993919-5a38eb00-51a7-11eb-87ba-4ddd5302a31a.gif)

**Comment by BladeMight** [comment]
В https://github.com/BladeMight/Mahou/commit/7af3192cae5de04a0c8964737ffe11deaa71f010 исправил проблемы с форматом CF_HDROP(при копировании списка файлов в проводнике) и похожими, теперь работает как нельзя лучше:
![yaaa](https://user-images.githubusercontent.com/11931888/104018631-2ec7f780-51cb-11eb-86c0-54d9236e58a3.gif)

**Comment by nikitasA1NT** [comment]
Обновился, проблемы нет. Правда бывает на самом деле иногда если быстро нечаянно переключаю с помощью Mahou клавишу, а потом обратно с помощью него же переключаю, то вставляется внезапно текст, который был в буфере обмена.
Еще есть некоторые проблемы, которые лучше было бы показать с помощью гифок. Не подскажете каким софтом пользуетесь для записи gif?

**Comment by BladeMight** [comment]
> Не подскажете каким софтом пользуетесь для записи gif?

ScreenToGif с настройкой на 30 fps.

**Comment by nikitasA1NT** [comment]
Так, долго пришлось отсутствовать, забыл уже основную свою проблему. Щас смотрел, вроде терпимо, если не злоупотреблять переключением раскладки, то никаких багов не возникает со вставкой текста из буфера. Что-то если всплывет - отпишусь = )
